### PR TITLE
Deduplicate membership changes

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

**Comment by NegativeMjark** [comment]
Is there a reason to only deduplicate joins, or could this be used for all membership changes?

File: synapse/handlers/room_member.py [hunk 1]
```diff
@@ -183,6 +185,44 @@ def update_membership(
             third_party_signed=None,
             ratelimit=True,
     ):
+        if action == Membership.JOIN:
+            result = self.join_cache.get((room_id, target))
```
Comment (NegativeMjark) [inline synapse/handlers/room_member.py]: Should ``remote_room_hosts`` and ``third_party_signed`` be part of the request key here?

**Comment by erikjohnston** [comment]
@NegativeMjark PTAL

**Comment by NegativeMjark** [comment]
LGTM

File: synapse/util/async.py [hunk 2]
```diff
@@ -137,3 +139,43 @@ def _concurrently_execute_inner():
...
+        self.key_to_defer = {}
+
+    @defer.inlineCallbacks
+    def queue(self, key):
+        current_defer = self.key_to_defer.setdefault(key, None)
+
+        new_defer = defer.Deferred()
+        self.key_to_defer[key] = new_defer
+
+        def remove_if_current(_):
```
Comment (NegativeMjark) [inline synapse/util/async.py]: Might not want to throw away the thing passed to the callback.

**Comment by NegativeMjark** [comment]
Maybe add a few comments explaining how the thing is supposed to work inside the ``queue`` method.

File: synapse/util/async.py [hunk 3]
```diff
@@ -137,3 +139,43 @@ def _concurrently_execute_inner():
...
+        with (yield linearizer.queue("test_key")):
+            # do some work.
+
+    """
+    def __init__(self):
+        self.key_to_defer = {}
+
+    @defer.inlineCallbacks
+    def queue(self, key):
+        current_defer = self.key_to_defer.setdefault(key, None)
```
Comment (NegativeMjark) [inline synapse/util/async.py]: Why is this ``setdefault`` rather than ``get``?

File: synapse/util/async.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/util/async.py]: Oh, I see. This is only ever passed the ``None`` from ``_trigger_defer_manager``

Comment (NegativeMjark) [inline synapse/util/async.py]: Would it be clearer to merge the ``_trigger_defer_manager`` and the ``remove_if_current`` into one thing?

Comment (erikjohnston) [inline synapse/util/async.py]: Possibly, though I think having them split up makes it clearer what each thing is doing.

Comment (NegativeMjark) [inline synapse/util/async.py]: Do we have problems with log contexts?

File: synapse/util/async.py [hunk 4]
```diff
@@ -137,3 +139,51 @@ def _concurrently_execute_inner():
...
+        self.key_to_defer[key] = new_defer
+
+        def remove_if_current(_):
+            d = self.key_to_defer.get(key)
+            if d is new_defer:
+                self.key_to_defer.pop(key, None)
+
+        new_defer.addBoth(remove_if_current)
+
+        yield current_defer
```
Comment (NegativeMjark) [inline synapse/util/async.py]: Is this going to have the right log context when it resolves?

File: synapse/util/async.py (hunk 2 above)
Comment (NegativeMjark) [inline synapse/util/async.py]: I think having two functions and an ``addBoth`` is less clear in this case.

Comment (erikjohnston) [inline synapse/util/async.py]: Well, your example doesn't actually work, as we specifically don't hit the callback for the most current defer in the list.

I've changed it though I think it just makes things look more complicated.

File: synapse/util/async.py [hunk 5]
```diff
@@ -137,3 +141,47 @@ def _concurrently_execute_inner():
...
+        if current_defer:
+            yield preserve_context_over_deferred(current_defer)
+
+        @contextmanager
+        def _ctx_manager(d):
+            try:
+                yield
+            finally:
+                d.callback(None)
+                d = self.key_to_defer.get(key)
```
Comment (NegativeMjark) [inline synapse/util/async.py]: might want to use a different ``d`` to avoid confusion.

**Comment by NegativeMjark** [comment]
LGTM

**Pull request merged**